### PR TITLE
dispatch_source_create returns NULL but is marked up as nonnull

### PR DESCRIPTION
**Pull request by mwwa**
dispatch_source_create is marked up as non-null but can actually return
null in programming error issues. In order to maintain the nullabilty
contract, these issues will now cause the overlay to assert and crash
rather than return a null pointer out of a non-null call.

Resolves: <rdar://problem/39937177>

### CODE REVIEW

**Comment by mwwa** [comment]
@swift-ci please test

**Review by jrose-apple — APPROVED** [review]

File: stdlib/public/SDK/Dispatch/Dispatch.mm [hunk 1]
```diff
@@ -47,3 +48,10 @@ static void _dispatch_overlay_constructor() {
   }
 }
 
+extern "C" void
+_swift_dispatch_source_create_abort(void)
+{
+  swift::swift_reportError(0,
+      "dispatch_source_create returned NULL, aborting");
```
[the remark this answers is not in the document]
Comment (mwwa) [reply stdlib/public/SDK/Dispatch/Dispatch.mm]: @jrose-apple I'll pick a more appropriate error string and open a new radar to cover the latter. We probably can check for common mistakes but it's probably something to handle in a new change?

**Comment by mwwa** [comment]
@swift-ci please test

**Comment by mwwa** [comment]
@swift-ci please tesst

**Comment by mwwa** [comment]
@swift-ci please test

**Comment by mwwa** [comment]
@swift-ci please test

**Comment by mwwa** [comment]
@swift-ci please test

**Comment by mwwa** [comment]
@jrose-apple Ok, I think this actually passes building again now. I'm ok if this just lands in master but if you think this is worth closing into a 4.2 branch I can try and do that today?

**Comment by jrose-apple** [comment]
I think it's all right. We're getting to the point where taking more changes into 4.2 is risky, and since this is covering for user error either way and isn't a regression it's reasonable for it to take one more release to show up.

**Pull request merged**